Improve efficiency of applying policy map changes

pippolo84 opened the pull request:
Revamp of #28472 

Original PR description:

---

This moves the incremental policy map change applying into the "sync-policy-map" controller instead of performing the applying of the changes directly. EndpointManager.UpdatePolicyMaps is refactored to trigger the controller of each endpoint and then to wait for reconciliation to finish. Semantics with regards to when e.g. UpdatePolicyMaps returns are unchanged, e.g. a failed reconciliation attempt does not block. Compared to prior a failed incremental policy map update is now retried by the controller.

This will reduce the amount of overhead in processing of ToFQDN policies in relation to a DNS request/response. Prior to this each DNS response would have triggered call to EndpointManager.UpdatePolicyMaps which in turn spawned a goroutine per endpoint to apply the changes. With high rate of DNS requests this may lead to lots of endpoint lock contention and context switching overhead due to queueing of calls to ApplyPolicyMaps that are unnecessary since a prior call likely had already applied the pending changes.

---

Apart from rebasing, the main difference with the original PR is the introduction of the `channels.Merge` helper to wait on multiple channels, as needed in `Daemon.updateSelectors` to wait for both the policy map changes and the ipcache.

joamaki committed:
channels: Add channels package for common channel-related utilities

DoneChan: the "chan struct{}" pattern for signaling completion is used
all over the place and thus giving it a type alias can be beneficial in
type signatures to underline its usage.

ClosedDoneChan: when a "done channel" is used where nothing needs to be
done, a new channel needs to be created and immediately closed. This
pattern seems to repeat a lot. We can instead create a singleton
DoneChan and close it at init time.

Merge: merges a slice of channels and return a channel that closed when
all the merged channels are closed. It can be used to wait for
completion of multiple "done channels".

Signed-off-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -1407,37 +1407,65 @@ func (e *Endpoint) syncPolicyMapWithDump() error {
 
-	ctrlName := fmt.Sprintf("sync-policymap-%d", e.ID)
-	e.controllers.CreateController(ctrlName,
+	return e.controllers.UpdateController(
```
We should use TriggerController() here, not UpdateController()  (except on initialization of course). UpdateController can miss updates.

squeed replied:
This will require a bit of state and swapping channels. It won't be pretty :-(.

pippolo84 replied:
> We should use TriggerController() here, not UpdateController() (except on initialization of course). UpdateController can miss updates.

I'm not sure I'm getting why UpdateController() can miss updates, can you please elaborate more?

squeed replied:
https://github.com/cilium/cilium/issues/29015 -- UpdateController is not reliable.

christarazi commented on pkg/channels/channels.go:
```diff
@@ -0,0 +1,36 @@
+
+// Merge merges a slice of channels and returns a channel that is
+// closed when all the merged channels are closed.
+func Merge[T any](cs ...<-chan T) <-chan T {
```
I think it would be useful to call out that the number of channels passed in will result in the same number of goroutines + 1

pippolo84 replied:
Actually, since our use case is limited to merge "notification channels" (i.e: a `<-chan struct{}` on which we want to lock until it is closed), we can do this with a single goroutine. I'm gonna change this to:

```
func Merge(cs ...<-chan struct{}) <-chan struct{} {
	out := make(chan struct{})
	go func() {
		for _, c := range cs {
			<-c
		}
		close(out)
	}()
	return out
}
```

christarazi commented on daemon/cmd/fqdn.go:
```diff
@@ -92,29 +93,34 @@ func (d *Daemon) updateSelectors(ctx context.Context, selectors map[policyApi.FQ
 
 	// This releases the nameManager lock; at this point, it is safe for another fqdn update to proceed
-	return
+	return channels.Merge(policyMapsDone, ipCacheDone)
```
Goroutines are spawned here as per my previous comment. I'm trying to understand how this fits into the part of the commit msg which states:

> Refactor UpdatePolicyMaps to leverage the triggering behaviour and to
avoid spawning a goroutine per endpoint.

christarazi replied:
I see that the comment was specific to `UpdatePolicyMaps`. However, I think my general point still stands: it's not clear what the specific improvement over goroutines are given the implications of `channels.Merge()`. How much have we reduced goroutines, or are we still using the same amount?

pippolo84 replied:
Solved with the change explained in https://github.com/cilium/cilium/pull/30535#discussion_r1525060133

christarazi commented on pkg/endpoint/bpf.go:
```diff
@@ -1407,37 +1407,65 @@ func (e *Endpoint) syncPolicyMapWithDump() error {
+				if err, _ = e.updateNetworkPolicy(proxyWaitGroup); err != nil {
+					e.getLogger().WithError(err).Warn("failed to publish policy update to L7 proxy")
+				}
+				proxyWaitGroup.Wait()
```
Why did this get moved into the controller DoFunc? It wasn't there previously.

squeed replied:
Right, this hits at a very subtle concurrency change.

Previously, this was divided in to three phases:

1. Queue policy notifications via `selectorcache.UpdateFQDNSelector() / .UpdateIdentities()`
2. Wait for notifications to be complete. Then, go `ep.UpdatePolicyMaps()` for all endpoints. Queue updates for Envoy
3. Wait for Envoy to finish updating.

We do lose a bit of parallelism, as we can make progress with the the next (2) as soon as it is done, rather than waiting for Envoy to ack any updates. I don't have a good intuition for whether or not this is relevant. My instinct is no; updating policy maps is not particularly complicated, so we probably don't gain much by making progress on the next iteration while waiting for Envoy. However, it would be good to reason about this.

pippolo84 replied:
> Why did this get moved into the controller DoFunc? It wasn't there previously.

Refactored to use another controller to apply the policy map pending changes.

> My instinct is no; updating policy maps is not particularly complicated, so we probably don't gain much by making progress on the next iteration while waiting for Envoy. However, it would be good to reason about this.

My hunch is that your intuition is correct.
Moreover, using a single controller to apply policy map pending changes **and** push them to the L7 proxy makes it impossible to get a notification for the completion of (2) before executing (3).
For now, I kept this approach with the dedicated controller too, but if you have better ideas I'm happy to hear them.

christarazi reviewed (commented):
Thanks for the PR! The general approach LGTM, a few comments below.

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -1407,37 +1407,65 @@ func (e *Endpoint) syncPolicyMapWithDump() error {
 				}
-				return e.syncPolicyMapWithDump()
+
+				err := e.syncPolicyMapWithDump(ctx)
```
`syncPolicyMapsWithDump()` ignores the context argument...

squeed replied:
Another divergence -- before, `ep.ApplyPolicyMapChanges()` does **not** dump, it only applies deltas. That makes it much more efficient. The background controller does an expensive map dump to make sure there hasn't been divergence. This change now adds this dump to the FQDN-response critical path, which is very latency sensitive.

We really need to not do map dumps during FQDN responses. This probably means we need two controllers :-(.

pippolo84 replied:
Right, even if we have this check to limit the rate of full reconciliation with dump:

```
if time.Since(e.policyMapLastFullReconciliation) < option.Config.PolicyMapFullReconciliationInterval {
	return nil
}
```

It would not be nice to hit that in the FQDN path. Let's use a separate controller, then.

pippolo84 replied:
Created a dedicated controller for incremental updates.

bimmlerd reviewed (approved):
CODEOWNERS looks good to me

tklauser commented on pkg/channels/channels.go:
```diff
@@ -0,0 +1,36 @@
+	var wg sync.WaitGroup
+	wg.Add(len(cs))
+	for _, c := range cs {
+		c := c
```
nit: with Go 1.22 this is no longer needed, see https://go.dev/doc/go1.22#language and https://go.dev/wiki/LoopvarExperiment.

pippolo84 replied:
Good point, thanks! :+1:

pippolo84 committed:
endpoint: incremental policy map updates via controller

Use a dedicated per-endpoint ontroller (incremental-sync-policymap) to
apply pending changes to the endpoint policy map.

This allows to keep a fixed pool of goroutines, each managed by its
respective endpoint controller, ready to run again when an incremental
sync for the policy map should be carried out.

A channel is used to signal the caller that the controller executed at
least once. In case of errors, the controller will retry the operation.

A minimum retry wait time of 100 millis and a maximum of 1 minute.  We
want to avoid aggressive retrying when the policy map is full.

Refactor UpdatePolicyMaps to leverage the triggering behaviour and to
avoid spawning a goroutine per endpoint.

Finally, rate limit the logging in case of errors when applying policy
map pending changes.

Signed-off-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -1146,25 +1156,75 @@ func (e *Endpoint) addPolicyKey(keyToAdd policy.Key, entry policy.MapStateEntry,
+// updateIncrementalPolicyMapController creates or triggers a controller to reconcile
+// pending policy map changes.
+func (e *Endpoint) updateIncrementalPolicyMapController() channels.DoneChan {
+	ch := make(chan struct{})
```
Does this work? What happens if multiple callers get "batched" in to the same update? Won't one of the channels be lost?

I strongly strongly strongly recommend using TriggerController, not UpdateController.
